### DISCUSSION TITLE
[Feature Request] RAPatches Off-Ramp for Hack Authors

### DISCUSSION THREAD

**Discussion by cIymax** [RetroAchievements/RAWeb#2712] [Ideas]
This is the RAWeb counterpart of my current proposal for RA to add an RAPatches off-ramp for hack authors, unless one already exists.

1. I did not want to be the one bringing this up, given all the good work that has gone into RA (including RAPatches) in the time that I've been with RA.

1. But I'm still doing so, because I think it'd be better for RA to hear it from me than much later on through social media from another hacker who may not have any ties to RA.

1. The mentioned off-ramp allows hack authors to preemptively request for their hacks not to be cloned into the RAPatches repo on Github.  If cloning has already occurred, the off-ramp allows requesting for the cloned copy to be removed from RAPatches.

1. RA does not host roms.  This is because the act of achievement modding does not give RA the right to host the underlying roms.  By the same token, however, achievement modding on top of a hack of a rom does not give RA the right to host a copy of the hack (or of the rom).

1. Ideally, RAPatches should only be hosting hacks that have entered the public domain or for which RA has received explicit permission to host (on-ramp approach).  Otherwise, such hosting by RA violates the IP of the hack authors, in my opinion.

1. RHDN honors hacker requests to take down their hacks.  RHDN, until their recent effective closure, actually adopted a whitelist approach of requiring a hacking team member to make a submission, before a hack can even be hosted.

1. Even sites hosting prepatched roms, which I won't name here, honor hacker requests to take down their hacks.  (Feel free to DM me for more details on this.)

1. As it stands, it seems that the only way for a hacker not to have a hack cloned is to make the hack into a paid hack.  For this to be the only way to opt out of cloning into RAPatches would seem to be putting the cart before the horse.

1. Unlike Nintendo, hack authors are very unlikely to sue.  That having been said, considering the overlap between the achievement-modding and romhacking communities, it would be contrary to what RA stands for, in my opinion, to clone hacks without hacker permission, let alone to refuse to remove hacks even when explicitly requested by a hacker.

1. If RA determines that having such a whitelist approach for (opt in to) cloning into RAPatches is too cumbersome, then the least RA should offer is an off-ramp for a hack author to get a hack into a blacklist from (opt out of) being cloned into RAPatches, in my opinion.

1. To be clear, this proposal is not about letting a hacker prevent a hack from getting achievements.  This proposal is only about letting a hacker keep a hack from being cloned into RAPatches.

1. An example implementation would update a hack's set page to add a link or button for a hack author to opt out of cloning into RAPatches.  Some verification of the identity of the hack author could be had, to prevent false submissions.  If the request is deemed valid, a flag can be stored in a database to indicate that the hack author has opted out of RAPatches cloning, and the hack should not be cloned into RAPatches.  An additional feature can allow a hacker to opt (back) in to RAPatches cloning.  The flag's value can be conveyed somewhere on the set page to all users or only to users with certain permissions.

Thanks for your attention, and if you need anything else from me, just let me know.

**Comment by wescopeland** [comment]
This is probably an unpopular opinion, but I would honestly rather see all unlinked patches have their sets demoted from the site. 

IMO, patches that aren’t archived in RAPatches is bad for players, devs, and the platform itself. What happens 5 years from now when half of these off-ramped download URLs 404? A player will have to resort to either piracy or downloading the patch from a “non-approved” location. 

I understand a patch author wanting to protect their work. I would rather them be convinced that us archiving it and working with them to celebrate their art is a good thing, both for them and their audience. If it’s a question of analytics, what better analytics could there be than player progression data through an achievements set?

There is probably some nuance here that I’m missing. I’m definitely open to changing my view, as I’m sure there’s other information that I’m not considering.

**Reply by cIymax** [reply]
Understood.  Then I think that at the very least, what we could be doing is to refrain from putting up a link to any RAPatches cloned copy as long as the hack author is hosting a freely downloadable copy themselves or via RHDN, a private Discord server, or similar.

I also strongly suggest we do the same for free-as-in-beer homebrews: link to their homebrew on Itch.io or similar and not our copy unless their copy is down.


It seems that some explaining of the position of the hack author may be helpful, so let me add my two cents here.  The cloning of a public build of a hack impacts the download counts on sites like RHDN, and the download counts are important for discoverability (like on the App Store by way of analogy).  The cloning of a private build of a hack, that's hosted exclusively on a private Discord, impacts the growth of the Discord community for the hack.  This can eventually reduce the hack author's motivation to continue developing the hack, in which case the players also stand to suffer.


Not directly related to RAWeb but related to this topic nonetheless is the following rule governing Acceptable Social Behavior:

> Advertising products, services, or other Discord servers without approval

I believe that a private hack Discord is relevant enough to a hack set that advertising a private hack Discord as a source for downloading a hack, isn't what this rule was intended to prohibit.  Same goes for a private homebrew Discord in relation to a homebrew set.  So, it would be great if this rule could be revisited and potentially clarified.

**Reply by Hexadigital** [reply]
> I also strongly suggest we do the same for free-as-in-beer homebrews: link to their homebrew on Itch.io or similar and not our copy unless their copy is down.

I always link directly to the author's site / download page for freely available homebrew that I create sets for, but I've been finding it more and more important to mirror the game and link that as well in the forum posts. One in particular has gotten more than ten updates since I made the set - nothing that changes the actual contents of the game, just part of a multi-port update that results in a new hash (by changing the version number in-game) - and as a result, I'd have to either mirror it or constantly go through a cycle of "user reports it's broken, so I test and link the new hash".

**Reply by wescopeland** [reply]
> I think that at the very least, what we could be doing is to refrain from putting up a link to any RAPatches cloned copy as long as the hack author is hosting a freely downloadable copy themselves or via RHDN, a private Discord server, or similar.

This sounds like a healthy compromise to me. I would be fine with the primary download link being elsewhere, as long as a mirror is always archived on RAPatches. This would be enforced via form validation in the management app: if a non-RAPatches link is provided as the patch URL, a mirror URL to RAPatches must be provided to submit the form. The mirror would be an additional database column, and would only be visible from the player's perspective if and when the primary download URL went offline.

My main concern is we ultimately need to have some semi-permanent way of letting players get the patches. If the recent chaos over at RHDN is anything to go by, it shows a lot of these external sources are more ephemeral than they might seem.

**Comment by cIymax** [comment]
Linking to a recent post on RA because I think it provides helpful context for this discussion.  RAWeb was also brought up specifically.
https://retroachievements.org/viewtopic.php?t=6824&c=244930#244930

**Comment by televandalist** [comment]
Some of this was already said in that thread, but here we go:

I'll start by saying that I do not look at this as RA hogging clicks/traffic. I rarely even look at the analytics for the repo because I don't really care about that. I don't even care about hacks in general. I don't play them. My sole interest in all of this is in making sure that hashes that are linked to RetroAchievements entries are verified and are reproduceable with little to no trouble. That's it. This results in me spending an obscene amount of time tracking things down so other people don't have to.

I try to make RAPatches as convenient as possible by doing the following:
- Having a consistent naming scheme across the board.
- For older patches, making new patches using current No Intro/Redump ROMs as a base. It's very annoying when I get a patch and find out that it'll only work with some garbage GoodTools ROM from 15 years ago. Hell, there was one patch that listed a Darkwater rip as the suggested base. :laughing:
- Sticking to BPS and xDelta whenever possible to ensure the patch will only work with the correct base and to minimize the convoluted patching process that some hacks/translations use.

Other things of note:
- A decent chunk of patches in the repo were made by running the current No Intro/Redump bases against pre-patched ROMs. Why? Sometimes it's preferable to going on a wild goose chase trying to find a patch that, for all I know, doesn't even exist. Most Russian translations and a lot of Portuguese translations are frequently done this way.
- Almost every NES patch was re-made by me using [this process](https://github.com/RetroAchievements/RAPatches/wiki/Creating-New-NES-Patches).
- I occasionally include sub-directories in RAPatches downloads that include the original patch, documentation, and extras from the original patch download. Even if I didn't, nothing is stopping anyone from opening PRs that add this content.

***

> Ideally, RAPatches should only be hosting hacks that have entered the public domain or for which RA has received explicit permission to host (on-ramp approach). Otherwise, such hosting by RA violates the IP of the hack authors, in my opinion.

Pardon my ignorance, but what constitutes a hack, which is a derivative work of actual copyrighted content no matter how you slice it, entering the public domain?

> RHDN honors hacker requests to take down their hacks. RHDN, until their recent effective closure, actually adopted a whitelist approach of requiring a hacking team member to make a submission, before a hack can even be hosted.

I will absolutely comply with taking patches off the repo so long as the hashes are unlinked from RA as well. The **only** case of this happening, at least that I can think of, is a hack author wanting their hacks removed from the site because they were leaving the hobby behind. I complied by removing the patches from the repo, but the dev for the sets left the hashes linked, so I removed the author's name from the patches and added them back. If hashes are only obtainable by joining other communities or through a paywall, then those hashes will likely not be linked; therefore, there would be no reason to add them to RAPatches. You yourself have gotten multiple messages from me regarding hashes you've linked. Most of the time, you told me they could be unlinked, which they were... and they were never brought up again. If I were trying to turn RAPatches into some almighty one-stop-shop for every patch ever, then that would have gone differently.

***

As it is now, you are basically saying that we should rely on other people, most of which we've had no interactions with, to make sure specific versions of patches remain available. If they decide to call it quits one day and get rid of everything (they're a moody bunch from what I've seen, so that seems entirely possible) or if a romhacking site gets bombarded with DDoS attacks, gets shut down, etc., then anyone wanting to get achievements in their hacks are screwed unless the patches are available on RAPatches. 

Anyway, there is definitely a conflict of interest here, but I assure you that mine starts and ends with the hashes. The majority of achievement developers don't bother with hash/patch management so most of it ends up falling on me. Hell, the RAPatches repo was originally mine, but I transferred ownership to RA after the ball got rolling. While I still maintain the repo and do the majority of the work, I don't have some authoritarian rule over it. I will compromise when/if necessary.

For what it's worth, RAPatches was initially meant to be a back-up. On the hash management pages, there are fields for "Resource Page URL" intended for links to No Intro's DAT-O-Matic, Redump, romhacking.net, SMWCentral, etc. These are not publicly visible, but I think they should be. Also, I think these links should take priority so links to RAPatches are treated as mirrors. However, that is out of my hands. 

<img width="603" alt="h" src="https://github.com/user-attachments/assets/f17c6407-4a3a-4942-9a7b-1a01667a6538">

**Reply by cIymax** [reply]
Thank you for your message and for all your hard work on the RAPatches repo.  As you know, I responded to you privately not long after your message was posted.  But I also wanted to tie up any loose ends in this public thread.

From the earlier discussion prior to your message, focus already shifted from the cloning of patches to the visibility of the cloned links.  In other words, in the spirit of compromise, the cloning is no longer at issue here.  I ask that until the visibility issue on RAWeb is corrected (an issue you said you were surprised to learn about and not responsible for), RA honors a hack author's explicit request, on a per-patch basis, to _not_ clone into the repo or to manually hide the link to a clone.  I will be indicating this in my RA official forum topic post for certain patches.

In closing, using the RPGe translation patch as an example, even Square Enix would _not_ have the right to distribute that translation patch.  More precisely, the same applies to the RPGe translation script itself, even if it infringes on their IP, _let alone_ to the RPGe code for handling variable-width font, which does _not_ infringe on their IP.  And if even Square Enix cannot legally distribute the RPGe translation patch, then neither can Retro Achievements via the RAPatches repo do so.  A double standard should never exist for game studios versus modders in terms of how we treat their work, period.

**Reply by wescopeland** [reply]
https://github.com/RetroAchievements/RAWeb/pull/3011

**Reply by cIymax** [reply]
I'd like to imagine that your NY's Day is better spent than on my silly request.  But in all seriousness, thank you, Wes, and Happy NY!
